pipeline/pipeline/pipeline.py

```diff
@@ -92,6 +92,12 @@ def read_input_group(self, **kwargs):
         return rg
 
     def write_output(self, resource, dest):  # pylint: disable=R0201
+        assert isinstance(resource, Resource)
```

Comment:
Can this ever be something besides Resource?

In short, do we need asserts for non-operational errors (something the user can't trigger)? I think that if the error isn't operational (user input), the assert is unnecessary (since it's a runtime check of a programmatic error), and should be written as a test. This applies also to several `if isinstance(...) else assert`

Reply:
This is a user facing method. I'm slowly converting all non-user methods to start with '_'. I'll make the error message nicer.

Reply:
Do we need the assert?